wolfcrypt/src/asn.c

```diff
@@ -4622,6 +4622,7 @@ int DsaPrivateKeyDecode(const byte* input, word32* inOutIdx, DsaKey* key,
         GetInt(&key->g, input, inOutIdx, inSz) < 0 ||
         GetOctetString(input, inOutIdx, &length, inSz) < 0 ||
         GetInt(&key->y, input, inOutIdx, inSz) < 0) {
+            wc_FreeDsaKey(key);
```

Comment:
Why do the free here and not the other places in this function that return? It seems like it would be the callers responsibility to free as it wasn't initialized in this function.

Reply:
@dgarske, Good question.  
wc_DsaPrivateKeyDecode() decodes a DER formatted certificate buffer containing DSA and it tries to decode two times if the first one fails. The default one can handle ASN.1 format with  ANSI X9.57 and I'm guessing the second one is an old format. 

In the default case,  Memory for key->p, q, g are allocated in this function. For example,  via a call trace wc_DsaPrivateKeyDecode() -> GetInt() -> mp_read_unsigned_bin() -> mp_grow() -> wolfSSL_Realloc() -> realloc() 

However, GetOctetString() returns ASN_PARSE_E because "/certs/dsa2048.der" is missing BIT STRING in the ASN.1 tag.

It tries decoding again with old format except wc_DsaPrivateKeyDecode() -> GetInt() -> mp_init() reinitalizes the old pointer  "a->dp = NULL" causing a memory leak since we now lost the previoius three pointers. 

The  wc_FreeDsaKey(key) ensures memory allocated in this function is freed properly. And the other places don't really need it.

Reply:
Hey Tesfa,

I see what's happening there now with trying an alternate path. Instead of calling `wc_FreeDsaKey` perhaps a better idea would be to explicitly clear/free each of the mp_int's.

How about doing the following instead:

```c
mp_clear(&key->x);
mp_clear(&key->y);
mp_clear(&key->g);
mp_clear(&key->q);
mp_clear(&key->p);
```

Thanks

Reply:
@dgarske,
Why call mp_clear() when wc_FreeDsaKey() does the same thing plus zeros the private key? 

We could potentially get rid of the alternative path altogether if "/certs/dsa2048.der" was created for testing in ad hoc form and can be updated to recent form. 

For A, I don't fully understand why ASN.1 tag of "/certs/dsa2048.der" shows :

SEQUENCE (6 elem)
  INTEGER 0
  INTEGER (2048 bit) 258230254080194213337757217925907840957462561833617845956684242041150…
  INTEGER (160 bit) 1322461528699410042723068051096948711363102409371
  INTEGER (2048 bit) 195453873072563761801372035426815536762132378374194145184630391395106…
  INTEGER (2047 bit) 129136525753911503723980884552023229069502235573611078157362960879415…
  INTEGER (160 bit) 873090317931674828959314930453392383793785105818

Reply:
You are making an assumption that no one ever adds any code to the wc_FreeDsaKey function. For example let’s say we add QAT support for DSA and the free function is called we will have an edge case big here. I’d prefer the inline mp_clear to ensure not future issues.

Reply:
@dgarske, Thanks for looking at this closely. I could be missing something but I don't see how your example of adding QAT support will lead to an edge case. 
For the sake of discussion, suppose you added a QAT support, I would think you need to properly initialize it in wc_InitDsaKey() (e.g. key->qat = malloc() or set it to NULL) and wc_FreeDsaKey() would check for NULL before freeing it.
So how exactly does this become an issue in the future?

In our test case, GetOctetString() returns ASN_PARSE_E but there could be cases where one of the GetInt() fails with other certificates.  I believe having wc_FreeDsaKey() is safe. It is one call and it zeros out the private key and frees the DSA elements and continues for the alternate pass. If there's is still an issue with the second path, it is left for the caller to handle the free.  I just don't see the advantage of the inline mp_clear() calls. Let me know if I missed something.